### PR TITLE
add setSync function to write user data in sync

### PR DESCRIPTION
**Pull request by damini-vashishtha**

### CODE REVIEW

File: lib/storage.js [hunk 1]
```diff
@@ -193,7 +193,7 @@ exports.get = function(key, options, callback) {
     function(object, callback) {
       var objectJSON = {};
       try {
-        objectJSON = JSON.parse(object);
+         objectJSON = JSON.parse(object);
```
Comment (jviotti) [inline lib/storage.js:196]: I think this change was accidental

File: lib/storage.js [hunk 2]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+ * @public
+ *
+ * @param {String} key - key
+ * @param {Object} json - json object
+ * @param {Function} callback - callback (error)
+ *
+ * @example
+ * const storage = require('electron-json-storage');
+ *
+ * storage.setSync('foobar', { foo: 'bar' }, function(error) {
```
Comment (jviotti) [inline lib/storage.js:456]: The example should not have a callback if its synchronous

File: lib/storage.js [hunk 3]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+    const data = JSON.stringify(json);
+
+    if (!data) {
+      return callback(new Error('Invalid JSON data'));
+    }
+
+    try {
+      mkdirp.sync(path.dirname(fileName));
+      fs.writeFileSync(fileName, data);
+      return callback();
```
Comment (jviotti) [inline lib/storage.js:473]: No need to have `callback` and call it if the call is synchronous. We can just `return` and `throw` instead

File: lib/storage.js [hunk 4]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+ *
+ * storage.setSync('foobar', { foo: 'bar' }, function(error) {
+ *   if (error) throw error;
+ * });
+ */
+exports.setSync = function(key, json, options, callback) {
+    const fileName = utils.getFileName(key, {
+      dataPath: options.dataPath
+    });
+    const data = JSON.stringify(json);
```
Comment (jviotti) [inline lib/storage.js:464]: Other functions take a `prettyPrinting` option. Here is how we use it in other functions:

```js
const data = JSON.stringify(json, null, (options.prettyPrinting ? 2 : 0));
```

File: lib/storage.js [hunk 5]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+
+    if (!data) {
+      return callback(new Error('Invalid JSON data'));
+    }
+
+    try {
+      mkdirp.sync(path.dirname(fileName));
+      fs.writeFileSync(fileName, data);
```
Comment (jviotti) [inline lib/storage.js:472]: As in the asynchronous `set` function, writing to the file directly is likely to result in data corruption from potential race conditions. Instead, we use `lock.lockSync` (see the `getSync`) function to acquire an exclusive lock (retrying the operation if the lock is taken).

Can we do this here too?

File: lib/storage.js [hunk 6]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+ *
+ * @example
+ * const storage = require('electron-json-storage');
+ *
+ * storage.setSync('foobar', { foo: 'bar' }, function(error) {
+ *   if (error) throw error;
+ * });
+ */
+exports.setSync = function(key, json, options, callback) {
+    const fileName = utils.getFileName(key, {
```
Comment (jviotti) [inline lib/storage.js:461]: I think indentation is off here. It should be two spaces based on the other contents of this file

File: lib/storage.js [hunk 7]
```diff
@@ -441,6 +441,42 @@ exports.set = function(key, json, options, callback, retries) {
...
+    if (!data) {
+      return callback(new Error('Invalid JSON data'));
+    }
+
+    try {
+      mkdirp.sync(path.dirname(fileName));
+      fs.writeFileSync(fileName, data);
+      return callback();
+    }
+    catch (error){
```
Comment (jviotti) [inline lib/storage.js:475]: There is a missing space between `)` and `{`

**Review by jviotti — CHANGES_REQUESTED** [review]
Looks promising! I left a few comments. Can you also add tests for it?

File: lib/storage.js (hunk 7 above)
Comment (damini-vashishtha) [reply lib/storage.js]: Done

File: lib/storage.js (hunk 6 above)
Comment (damini-vashishtha) [reply lib/storage.js]: Done

File: lib/storage.js (hunk 5 above)
Comment (damini-vashishtha) [reply lib/storage.js]: Done

File: lib/storage.js (hunk 4 above)
Comment (damini-vashishtha) [reply lib/storage.js]: Done

File: lib/storage.js (hunk 2 above)
Comment (damini-vashishtha) [reply lib/storage.js]: Done

**Comment by jviotti** [comment]
Awesome! Looking forward to some basic tests and I'll merge + release this

**Review by silentsakky — APPROVED** [review]

**Comment by damini-vashishtha** [comment]
@jviotti I am done with all the suggested changes. Can you please merge and release this asap?

**Comment by jviotti** [comment]
@damini-vashishtha The Travis CI required check is not firing and I don't have admin access on this repo to disable it myself. I'm reaching out to relevant people to get proper access. In the mean-time, I believe you can use npm to install the module directly from the git branch.

**Comment by jviotti** [comment]
Done! v4.6.0 has been published to npm

**Pull request merged**